File: frontend/ui-clusters/src/kui/ui/clusters/admin/ClusterForm.scala
package kui.ui.clusters.admin

import kui.cluster.contract.dto.{AdminTuningWrite, ClusterSecurityWrite, ClusterWriteRequest}
import kui.contracts.cluster.ClusterRowDto
import kui.kernel.Secret

/** What the operator has typed, as plain strings, and the rules for turning it into a request.
  *
  * ==Strings, not parsed values==
  *
  * Every field is a `String`, including the three numbers. A form that held an `Int` would have to decide
  * what an empty box or a half-typed `12x` *is* while the operator is still typing, and every answer to that
  * is wrong: zero is a number they did not write, and refusing the keystroke stops them deleting a digit.
  * Parsing happens once, in [[toRequest]], at the moment they ask for something to be done.
  *
  * ==What is deliberately not here==
  *
  * Truststore and keystore material. The write contract accepts both as base64 blobs and this form does not
  * offer them, which means TLS configured with a *private* certificate authority still has to be written into
  * the deployment's configuration file. That is a real gap and it is named here rather than papered over: a
  * textarea for a base64 keystore is not a usable way to give KUI a certificate, and the file upload that
  * would be is a piece of work in its own right. `SSL` and `SASL_SSL` against a cluster whose certificate the
  * JVM's default trust store already accepts — which is every managed service — work from this form today.
  */
final case class ClusterForm(
    name: String,
    bootstrapServers: String,
    readOnly: Boolean,
    protocol: String,
    mechanism: String,
    username: String,
    password: String,
    verifyHostname: Boolean,
    timeoutMs: String,
    batchSize: String,
    parallelism: String
) {

  /** Whether this protocol needs a mechanism, a username and a password. */
  def isSasl: Boolean = protocol == ClusterForm.SaslPlaintext || protocol == ClusterForm.SaslSsl

  /** Whether hostname verification is a question at all. It is a TLS setting; on a plaintext connection there
    * is no certificate to verify a hostname against, and showing the control would suggest otherwise.
    */
  def isTls: Boolean = protocol == ClusterForm.Ssl || protocol == ClusterForm.SaslSsl

  /** The request, or every reason this form is not one yet.
    *
    * Every reason, accumulated, and not the first: someone who got three fields wrong should be told about
    * all three rather than discovering them one save at a time. The server accumulates for the same reason
    * (ADR-013) and refuses the same things; these checks exist to answer *before* a round trip and are
    * deliberately a subset — the server remains the authority, and this form never claims something is valid
    * that the server then rejects.
    *
    * The password is the one field this cannot check on an edit. KUI never sends a stored credential back to
    * the browser, so an edit form starts with the box empty, and an empty box means "leave the password
    * alone" — which this cannot express, because the write contract has no such value. See
    * [[ClusterForm.passwordWarning]] for what the screen says about it instead.
    */
  def toRequest: Either[List[String], ClusterWriteRequest] = {
    val problems = List.newBuilder[String]

    if name.trim.isEmpty then problems += "A name is required."
    if bootstrapServers.trim.isEmpty then problems += "At least one broker address is required."

    if isSasl && mechanism.trim.isEmpty then problems += "A SASL connection needs a mechanism."
    if isSasl && username.trim.isEmpty then problems += "A SASL connection needs a username."
    if isSasl && password.isEmpty then problems += "A SASL connection needs a password."

    val timeout = timeoutMs.trim.toLongOption
    val batch = batchSize.trim.toIntOption
    val threads = parallelism.trim.toIntOption

    if timeout.isEmpty then problems += "The admin timeout must be a whole number of milliseconds."
    if batch.isEmpty then problems += "The batch size must be a whole number."
    if threads.isEmpty then problems += "The parallelism must be a whole number."

    val found = problems.result()

    (found, timeout, batch, threads) match {
      case (Nil, Some(timeoutValue), Some(batchValue), Some(threadValue)) =>
        Right(
          ClusterWriteRequest(
            name = name.trim,
            readOnly = readOnly,
            bootstrapServers = bootstrapServers.trim,
            security = ClusterSecurityWrite(
              protocol = protocol,
              mechanism = Option.when(isSasl)(mechanism.trim),
              username = Option.when(isSasl)(username.trim),
              password = Option.when(isSasl)(Secret(password)),
              truststore = None,
              keystore = None,
              verifyHostname = verifyHostname
            ),
            properties = Map.empty,
            admin = AdminTuningWrite(timeoutValue, batchValue, threadValue)
          )
        )
      case _ => Left(found)
    }
  }
}

object ClusterForm {

  val Plaintext: String = "PLAINTEXT"
  val Ssl: String = "SSL"
  val SaslPlaintext: String = "SASL_PLAINTEXT"
  val SaslSsl: String = "SASL_SSL"

  val Protocols: List[(String, String)] =
    List(
      Plaintext -> "PLAINTEXT — no encryption, no authentication",
      Ssl -> "SSL — encrypted, no authentication",
      SaslPlaintext -> "SASL_PLAINTEXT — authenticated, not encrypted",
      SaslSsl -> "SASL_SSL — authenticated and encrypted"
    )

  /** The three mechanisms KUI is integration-tested against a real broker with.
    *
    * The vendor mechanisms — `AWS_MSK_IAM`, `OAUTHBEARER` and the rest — are configuration-file only, and the
    * write endpoint refuses them by name rather than accepting a value it cannot exercise. Offering them here
    * would produce a form whose Save is always refused.
    */
  val Mechanisms: List[(String, String)] =
    List(
      "PLAIN" -> "PLAIN",
      "SCRAM-SHA-256" -> "SCRAM-SHA-256",
      "SCRAM-SHA-512" -> "SCRAM-SHA-512"
    )

  /** An empty form, with the defaults a new cluster starts from.
    *
    * The three admin numbers are the domain's own defaults spelled out rather than left blank, because a
    * blank timeout is not a smaller decision than a wrong one — it is the same decision made by whoever wrote
    * the fallback, invisibly.
    */
  val Empty: ClusterForm =
    ClusterForm(
      name = "",
      bootstrapServers = "",
      readOnly = false,
      protocol = Plaintext,
      mechanism = "SCRAM-SHA-512",
      username = "",
      password = "",
      verifyHostname = true,
      timeoutMs = "60000",
      batchSize = "200",
      parallelism = "4"
    )

  /** The form for editing a cluster the dashboard already shows.
    *
    * Everything except the credentials, because KUI never sends a stored credential back to the browser — the
    * read model has a `truststoreConfigured` boolean and no password field, by design. The admin numbers are
    * not on the read model either, so they come back as the defaults; the screen says so.
    */
  def of(row: ClusterRowDto): ClusterForm =
    Empty.copy(
      name = row.name,
      bootstrapServers = row.bootstrapServers,
      readOnly = row.readOnly,
      protocol = row.security.protocol,
      mechanism = row.security.mechanism.getOrElse(Empty.mechanism)
    )

  /** What an edit form has to say about the password box being empty.
    *
    * The honest sentence, because the alternative — showing dots and pretending there is a value — would make
    * an operator who did not touch the field wipe the credential by saving. The write endpoint has no "leave
    * it as it was" value: a `PUT` replaces the record.
    */
  val passwordWarning: String =
    "KUI never sends a stored password back to the browser, so this box starts empty. Saving replaces " +
      "the whole cluster definition, so type the password again even if it has not changed."

  /** And what it has to say about the three admin numbers, for the same reason. */
  val tuningWarning: String =
    "The admin timeouts are not on the read model, so these show the defaults rather than what is stored. " +
      "Saving writes whatever is in these boxes."
}

File: frontend/packages/feature-clusters/src/clusterForm.ts
/**
 * The form behind "add a cluster", "change one" and "remove one", and the rules it enforces.
 *
 * ## What this form deliberately cannot do
 *
 * It cannot supply a truststore or a keystore. Those are configuration-file only, and that is a real
 * gap named here rather than papered over: a textarea for a base64 keystore is not a usable way to
 * give KUI a certificate, and the file upload that would be is a piece of work in its own right.
 * `SSL` and `SASL_SSL` against a cluster whose certificate the JVM's default trust store already
 * accepts — which is every managed service — work from this form today.
 *
 * ## Every reason, not the first one
 *
 * Somebody who got three fields wrong should be told about all three rather than discovering them
 * one save at a time. The server accumulates for the same reason (ADR-013) and refuses the same
 * things; these checks exist to answer *before* a round trip and are deliberately a subset — the
 * form never claims something is valid that the server then rejects.
 */

export const PLAINTEXT = "PLAINTEXT";
export const SSL = "SSL";
export const SASL_PLAINTEXT = "SASL_PLAINTEXT";
export const SASL_SSL = "SASL_SSL";

export const PROTOCOLS: readonly { readonly value: string; readonly label: string }[] = [
  { value: PLAINTEXT, label: "PLAINTEXT — no encryption, no authentication" },
  { value: SSL, label: "SSL — encrypted, no authentication" },
  { value: SASL_PLAINTEXT, label: "SASL_PLAINTEXT — authenticated, not encrypted" },
  { value: SASL_SSL, label: "SASL_SSL — authenticated and encrypted" },
];

/**
 * The three mechanisms KUI is integration-tested against a real broker with.
 *
 * The vendor mechanisms — `AWS_MSK_IAM`, `OAUTHBEARER` and the rest — are configuration-file only,
 * and the write endpoint refuses them by name rather than accepting a value it cannot exercise.
 * Offering them here would produce a form whose Save is always refused.
 */
export const MECHANISMS: readonly { readonly value: string; readonly label: string }[] = [
  { value: "PLAIN", label: "PLAIN" },
  { value: "SCRAM-SHA-256", label: "SCRAM-SHA-256" },
  { value: "SCRAM-SHA-512", label: "SCRAM-SHA-512" },
];

export interface ClusterForm {
  /**
   * The slug this cluster is addressed by: it is in every URL and in every audit line, and it cannot
   * be changed afterwards without the links people have kept stopping working. Collected only when
   * adding — an edit addresses an existing record and must not move it.
   */
  readonly id: string;
  readonly name: string;
  readonly bootstrapServers: string;
  readonly readOnly: boolean;
  readonly protocol: string;
  readonly mechanism: string;
  readonly username: string;
  readonly password: string;
  readonly verifyHostname: boolean;
  readonly timeoutMs: string;
  readonly batchSize: string;
  readonly parallelism: string;
}

/**
 * An empty form, with the defaults a new cluster starts from.
 *
 * The three admin numbers are the domain's own defaults spelled out rather than left blank, because
 * a blank timeout is not a smaller decision than a wrong one — it is the same decision, made by
 * whoever wrote the fallback, invisibly.
 */
export const EMPTY_CLUSTER_FORM: ClusterForm = {
  id: "",
  name: "",
  bootstrapServers: "",
  readOnly: false,
  protocol: PLAINTEXT,
  mechanism: "SCRAM-SHA-512",
  username: "",
  password: "",
  verifyHostname: true,
  timeoutMs: "30000",
  batchSize: "200",
  parallelism: "8",
};

/** Whether this protocol needs a mechanism, a username and a password. */
export function isSasl(form: ClusterForm): boolean {
  return form.protocol === SASL_PLAINTEXT || form.protocol === SASL_SSL;
}

/**
 * Whether hostname verification is a question at all.
 *
 * It is a TLS setting: on a plaintext connection there is no certificate to verify a hostname
 * against, and showing the control would suggest otherwise.
 */
export function isTls(form: ClusterForm): boolean {
  return form.protocol === SSL || form.protocol === SASL_SSL;
}

/** The server's own rule for a cluster id: `^[a-z0-9]([a-z0-9-]{0,62}[a-z0-9])?$`. */
const ID_PATTERN = /^[a-z0-9]([a-z0-9-]{0,62}[a-z0-9])?$/;

/**
 * A name turned into a plausible id, for the add form to start from.
 *
 * A suggestion and not a rule: the box stays editable, because the id is in every URL and every
 * audit line for the life of the cluster and the person adding it may well want a shorter one than
 * "our-staging-kafka-eu-west-1".
 */
export function suggestId(name: string): string {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, 64)
    .replace(/-+$/, "");
}

/** A whole number, or `undefined`. `Number("")` is 0 and `Number("8x")` is NaN; neither is a count. */
function whole(text: string): number | undefined {
  const trimmed = text.trim();
  if (trimmed === "") return undefined;
  const value = Number(trimmed);
  return Number.isInteger(value) ? value : undefined;
}

/** The write request this form describes, or every reason it is not one yet. */
export function toRequest(
  form: ClusterForm,
):
  | { readonly ok: true; readonly request: Record<string, unknown> }
  | { readonly ok: false; readonly problems: readonly string[] } {
  const problems: string[] = [];

  if (form.name.trim() === "") problems.push("A name is required.");
  if (form.id.trim() === "") {
    problems.push("An id is required.");
  } else if (!ID_PATTERN.test(form.id.trim())) {
    // The server's own rule, quoted. It is in every URL this cluster appears in, so the constraint
    // is not arbitrary tidiness — a slug with a slash or a space in it is a path that does not route.
    problems.push(
      "An id is lowercase letters, digits and hyphens, starting and ending with a letter or digit.",
    );
  }
  if (form.bootstrapServers.trim() === "")
    problems.push("At least one broker address is required.");

  if (isSasl(form)) {
    if (form.mechanism.trim() === "") problems.push("A SASL connection needs a mechanism.");
    if (form.username.trim() === "") problems.push("A SASL connection needs a username.");
    /*
     * The one field this cannot check on an edit. KUI never sends a stored credential back to the
     * browser, so an edit form starts with the box empty — and an empty box means "leave the
     * password alone", which the write contract has no value for: a `PUT` replaces the record. See
     * `PASSWORD_WARNING` for what the screen says about it instead.
     */
    if (form.password === "") problems.push("A SASL connection needs a password.");
  }

  const timeout = whole(form.timeoutMs);
  const batch = whole(form.batchSize);
  const threads = whole(form.parallelism);

  if (timeout === undefined)
    problems.push("The admin timeout must be a whole number of milliseconds.");
  if (batch === undefined) problems.push("The batch size must be a whole number.");
  if (threads === undefined) problems.push("The parallelism must be a whole number.");

  if (
    problems.length > 0 ||
    timeout === undefined ||
    batch === undefined ||
    threads === undefined
  ) {
    return { ok: false, problems };
  }

  return {
    ok: true,
    request: {
      name: form.name.trim(),
      readOnly: form.readOnly,
      bootstrapServers: form.bootstrapServers.trim(),
      security: {
        protocol: form.protocol,
        // Omitted rather than sent empty on a non-SASL connection: a mechanism on a PLAINTEXT
        // cluster is a setting that cannot apply, and the server refuses it.
        ...(isSasl(form)
          ? {
              mechanism: form.mechanism.trim(),
              username: form.username.trim(),
              password: form.password,
            }
          : {}),
        verifyHostname: form.verifyHostname,
      },
      properties: {},
      admin: { timeoutMs: timeout, batchSize: batch, parallelism: threads },
    },
  };
}

/**
 * What the screen says about the password box on an edit.
 *
 * The honest sentence, because the alternative — showing dots and pretending there is a value —
 * would let an operator who did not touch the field wipe the credential by saving.
 */
export const PASSWORD_WARNING =
  "KUI never sends a stored password back to the browser, so this box starts empty. Saving replaces " +
  "the whole cluster definition, so type the password again even if it has not changed.";

/** And what it says about the three admin numbers, for the same reason. */
export const TUNING_WARNING =
  "The admin timeouts are not on the read model, so these show the defaults rather than what is " +
  "stored. Saving writes whatever is in these boxes.";

/** The form for an existing cluster: everything the read model carries, and defaults for the rest. */
export function formFor(row: {
  readonly id: string;
  readonly name: string;
  readonly bootstrapServers: string;
  readonly readOnly: boolean;
  readonly security?:
    { readonly protocol?: string; readonly mechanism?: string | null } | undefined;
}): ClusterForm {
  return {
    ...EMPTY_CLUSTER_FORM,
    id: row.id,
    name: row.name,
    bootstrapServers: row.bootstrapServers,
    readOnly: row.readOnly,
    protocol: row.security?.protocol ?? PLAINTEXT,
    mechanism: row.security?.mechanism ?? EMPTY_CLUSTER_FORM.mechanism,
  };
}
